feat(ci): Added initial build and upload workflow

MoritzThomasHuebner opened the pull request:
Relating to #13450

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->
Created in pairing with @alexzurbonsen 

## Summary
- Workflow builds nettle into a .deb file using fpm
- Workflow installs nettle with the .deb to test if the .deb works
- Extra check to see if the libnettle.so file exists
- Upload artifact to github
- Upload to artifactory
- Access is currently restricted to a certain set of users.

The actor can configure:
- The repository (default debian-test)
- The distribution (default focal-ci)
- Whether to do a try run, i.e. not actually upload the .deb (default true)

The magma repository probably still needs an access token for the artifactory. Currently, authentication works via username/password secrets. 

<!-- Enumerate changes you made and why you made them -->

## Test Plan
Manually tested using the dry-run option. Also tested with a temporary folder on the artifactory (janheidbrink) with a temporary token:
https://artifactory.magmacore.org/ui/repos/tree/General/janheidbrink%2Fpool%2Ffocal-ci%2Foai-nettle_2.5-1ubuntu20.04_amd64.deb

<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

Neudrino commented on .github/workflows/build_magma_dep.yml:
```diff
@@ -0,0 +1,73 @@
+jobs:
+  build_deps:
+    runs-on: ubuntu-20.04
+    if: contains('["maxhbr","nstng", "Neudrino", "tmdzk", "alexzurbonsen", "MoritzThomasHuebner"]', github.actor)
```
_nitpick:_ inconsistent spacing here.

MoritzThomasHuebner replied:
Fixed

Neudrino commented on .github/workflows/build_magma_dep.yml:
```diff
@@ -0,0 +1,73 @@
+    runs-on: ubuntu-20.04
+    if: contains('["maxhbr","nstng", "Neudrino", "tmdzk", "alexzurbonsen", "MoritzThomasHuebner"]', github.actor)
+    steps:
+      - uses: actions/checkout@7884fcad6b5d53d10323aee724dc68d8b9096a2e # pin@v2
```
As this is a new workflow, can we us the newest v3 immediately, please.

MoritzThomasHuebner replied:
:+1:

Neudrino replied:
Sorry, should have made it clear though, that we should still pin the version to an exact hash. :innocent:

MoritzThomasHuebner replied:
All good, pinning versions is not consistent across all workflows, so I was unsure.

Neudrino replied:
Since https://github.com/magma/magma/pull/12990 was merged, all Github actions should be pinned to hashes.

Neudrino replied:
But I am confused now, the comment says it is version 2 again.
We should have it pinned to the hash of the latest version 3 though.

MoritzThomasHuebner replied:
Sorry, that was my bad. The comment should now say it's version 3 and the git commit hash should be 2541b1294d2704b0964813337f33b291d3f8596b

Neudrino commented on .github/workflows/build_magma_dep.yml:
```diff
@@ -0,0 +1,73 @@
+    steps:
+      - uses: actions/checkout@7884fcad6b5d53d10323aee724dc68d8b9096a2e # pin@v2
+        with:
+          fetch-depth: 0
```
Why would that be necessary?
We should not be doing anything on foreign commits in this workflow.

MoritzThomasHuebner replied:
I removed this

Neudrino commented on .github/workflows/build_magma_dep.yml:
```diff
@@ -0,0 +1,73 @@
+      - name: Build nettle
+        shell: bash
+        run: |
+          mkdir ~/magma-packages
```
- In scripts always use mkdir with the `-p` option
- Is it safe to use the home dir on a Github runner?

MoritzThomasHuebner replied:
- I added the `-p` on the `mkdir` command.
- By using the home directory do you mean the use of `~` or just being in `/home/runner/` in general?

Neudrino replied:
`home dir`: The latter. Using `~` should be fine. However, I was not sure, if and how home directory would be usable on those runners. But if it works, it is probably fine.

Neudrino commented on .github/workflows/build_magma_dep.yml:
```diff
@@ -0,0 +1,73 @@
+      - name: Testing if .deb file can be installed
+        shell: bash
+        run: |
+          sudo apt install -y ~/magma-packages/*.deb
```
This is the third time `~/magma-packages` is used. We should really use an `env` for it than.

MoritzThomasHuebner replied:
I created the env

Neudrino commented on .github/workflows/build_magma_dep.yml:
```diff
@@ -0,0 +1,73 @@
+      - name: Testing is libnettle.so exists.
+        shell: bash
+        run: |
+          FILE=/usr/lib/libnettle.so
```
Consider using a [Github env](https://docs.github.com/en/actions/learn-github-actions/environment-variables) instead.

MoritzThomasHuebner replied:
I created an env

Neudrino commented on .github/workflows/build_magma_dep.yml:
```diff
@@ -0,0 +1,73 @@
+      - name: Dry run - publish debian packages
+        if: ${{ github.event.inputs.dry_run == 'true'}}
+        run: |
+          jf rt u --recursive=false --url=https://artifactory.magmacore.org:443/artifactory/ --detailed-summary \
```
- Every command option should have its own line.
- We might be able to get rid of the URL by [including it into the token](https://github.com/marketplace/actions/setup-jfrog-cli#creating-the-configuration-on-your-local-machine)?

MoritzThomasHuebner replied:
- I reformatted so that every command has its own line
- It seems certainly possible, however this would be hard for me to test since I don't have access to the magma github secrets. Other workflows also state the artifactory URL explicitly.

Neudrino commented on .github/workflows/build_magma_dep.yml:
```diff
@@ -0,0 +1,73 @@
+        run: |
+          jf rt u --recursive=false --url=https://artifactory.magmacore.org:443/artifactory/ --detailed-summary \
+            --target-props="deb.distribution=${{ github.event.inputs.distribution }};deb.component=main;deb.architecture=amd64" \
+            "~/magma-packages/(*).deb" ${{ github.event.inputs.repo }}/pool/${{ github.event.inputs.distribution }}/{1}.deb
```
One should probably avoid this superfluous duplication if the only difference is the `--dry-run` option.

A solution could involve [environment variables](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable).
Something similar to the following:
```yaml
steps:
  - name: Set the value
    id: is_dry
    if: ${{ github.event.inputs.dry_run == 'true' }}
    run: |
      echo "IS_DRY='--dry-run'" >> $GITHUB_ENV
  - name: Use the value
    id: publish
    run: |
      jf rt u --all_other_options_here \ 
        "${{ env.IS_DRY }}"
```

MoritzThomasHuebner replied:
I made the change as you suggested but it took me while to figure out what was going wrong. I had to remove some of the quotation marks to make it work.

MoritzThomasHuebner committed:
feat(ci): Added initial build and upload workflow

- Workflow builds nettle into a .deb file using fpm
- Workflow installs nettle with the .deb to test if the .deb works
- Extra check to see if the libnettle.so file exists
- Upload artifact to github
- Upload to artifactory

The actor can configure:
- The repository (default debian-test)
- The distribution (default focal-ci)
- Whether to do a try run, i.e. not actually upload the .deb (default true)

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>
